add docs to upgrade to Flux GA (#3717)
weaveworks/weave-gitops: website/docs/guides/fluxga-upgrade.mdx

---
title: Upgrade to Flux GA
hide_title: true
---

# Upgrade to Flux GA

We are very excited due Flux GA is being [released!](https://github.com/fluxcd/flux2/releases)

> This guide aims to answer some [common questions](#faq) before starting the upgrade, as well as, guiding how to do

Author: reworded 🙇

- `Flux GA` as the [latest Flux GA Release Candidate](https://github.com/fluxcd/flux2/releases/tag/v2.0.0-rc.3)
- `Weave Gitops` as the [latest Weave Gitops Enterprise release](https://github.com/weaveworks/weave-gitops-enterprise/releases/tag/v0.23.0)

## Important events during this period

_Dates for these events are not finalized and will depend on when the releases are made_

- The release of a version of Weave Gitops OSS/EE that will require Flux GA be installed (in the weeks after Flux GA is released).
- The release of a version of Flux GA that will drop support for `< v1` Custom Resources (~6-12 months after Flux GA is released).

You can take steps today to address both these events.

## FAQ

Here you could find the most common questions around upgrading.

### Why upgrade to Flux GA

Flux Beta APIs have been stable and used in production for quite some time now. However, with Flux GA there is big reason to move to v1:
it is the main supported api version for new features and developments. Features like [horizontal scaling](https://fluxcd.io/flux/cheatsheets/sharding/).

> are only available in Flux GA. Also have into consideration that beta apis will be removed after 6 months.

Reviewer: ```suggestion
are only available in Flux GA. Also, have into consideration that beta APIs will be removed after 6 months.
```

### Will be able to use Flux Beta once Flux GA is released?

It is encouraged you to plan upgrading to Flux GA using its release candidates train. However, if for reasons
you are not able to, you will be able to use Weave Gitops with Flux Beta, but we expect you could find increasing limitations
as new features would design with Flux GA APIs in mind.

## Upgrade

:::info
_Note: Flux GA release candidates have the same kubernetes compatibility as Flux 0.41.2_
:::

:::info Hosted flux?
If you are using a hosted Flux version, please check with your provider if they support Flux GA before upgrading following this guide.
Known hosted flux providers:

- EKS Anywhere
- [Azure AKS Flux-Gitops extension](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/extensions-release#flux-gitops)

As of writing they do not yet support the new version so please wait before upgrading to Flux GA.
:::

Below, we'll take you through the multiple steps required to migrate to your system to Flux GA. After each step the cluster will be
in a working state, so you can take your time to complete the migration.

1. Upgrade to [latest Weave Gitops](https://docs.gitops.weave.works/docs/releases/)

> 2. Upgrade to Flux GA in `ClusterBootstrapConfig`s

Author: applied

3. Upgrade to Flux GA on your leaf clusters and management clusters
4. Upgrade GitopsTemplates, GitopsSets and ClusterBootstrapConfigs

### 1. Upgrade to latest WGE

Use your regular WGE upgrade procedure to bring it to the [latest](https://docs.gitops.weave.works/docs/releases/)

:::warning upgrading to Flux GA with WGE v0.22.0?

Ensure that you address [this](#v0220)
:::

At this stage you have Weave Gitops running Flux Beta.

> ### 2. Upgrade to Flux GA in `ClusterBootstrapConfig`s

Reviewer: ```suggestion
### 2. Upgrade to Flux GA in ClusterBootstrapConfigs
```
to make consistent with point 4. below

First we ensure any new clusters are bootstrapped with Flux GA, then we'll upgrade the existing clusters.

`ClusterBootstrapConfig` will most often contain an invocation of `flux bootstrap`, make sure the image is using `v2`.

<details><summary>Expand to see an example</summary>

```patch
diff --git a/tools/dev-resources/user-guide/cluster-bootstrap-config.yaml b/tools/dev-resources/user-guide/cluster-bootstrap-config.yaml
index bd41ec036..1b21df860 100644
--- a/tools/dev-resources/user-guide/cluster-bootstrap-config.yaml
+++ b/tools/dev-resources/user-guide/cluster-bootstrap-config.yaml
@@ -1,34 +1,34 @@
 apiVersion: capi.weave.works/v1alpha1
 kind: ClusterBootstrapConfig
 metadata:
   name: capi-gitops
   namespace: default
 spec:
   clusterSelector:
     matchLabels:
       weave.works/capi: bootstrap
   jobTemplate:
     generateName: "run-gitops-{{ .ObjectMeta.Name }}"
     spec:
       containers:
-        - image: ghcr.io/fluxcd/flux-cli:v0.34.0
+        - image: ghcr.io/fluxcd/flux-cli:v2
           name: flux-bootstrap
           ...
```
</details>

> At this stage you new bootstrapped clusters will run Flux GA.

Reviewer 2: ```suggestion
At this stage, your new bootstrapped clusters will run Flux GA.
```

### 3. Upgrade to Flux GA on your existing leaf clusters and management clusters

> Follow the upgrade instructions from the [Flux v2.0.0-rc.1 release notes](https://github.com/fluxcd/flux2/releases/tag/v2.0.0-rc.1)

Reviewer 2: ```suggestion
Follow the upgrade instructions from the [Flux v2.0.0-rc.1 release notes](https://github.com/fluxcd/flux2/releases/tag/v2.0.0-rc.1).
```

At minimum, you'll need to rerun the `flux bootstrap` command on your leaf clusters and management clusters.

> You'll also need to bump api versions in your manifests to `v1` as described in the flux upgrade instructions:

Reviewer 2: ```suggestion
You'll also need to bump APIs version in your manifests to `v1` as described in the flux upgrade instructions:
```

### 4. Upgrade GitopsTemplates, GitopsSets and ClusterBootstrapConfigs

Bumping the APIs version in manifests can be done gradually. It is advised to not delay this procedure as the beta versions will be removed after 6 months.

#### `GitopsTemplate` and `CAPITemplate`

Update `GitRepository` and `Kustomization` CRs in the `spec.resourcetemplates` to `v1` as described in the flux upgrade instructions.

#### `GitopsSets`

Update `GitRepository` and `Kustomization` CRs in the `spec.template` of your `GitopsSet` resources to `v1` as described in the flux upgrade instructions.

### 5. Future steps

At this stage all clusters are running Flux GA. This will allow you to seamlessly upgrade to the upcoming release of Weave Gitops that will drop support for Flux Beta.

If you have updated your `Kustomization` and `GitRepository` resources to `v1` you will also be able to upgrade to the future release of Flux that will drop support for `< v1` apis.

:::info

Weave Gitops Enterprise 0.23.0 generates v1beta2 `Kustomization` Custom Resources when using the _Add app_ and _Cluster Template_ features.

This is to support installations still using Flux Beta.

> After upgrading to the release of Weave Gitops that drops support for Flux Beta Weave Gitops Enterprise will create `v1` Resources.

Reviewer 2: ```suggestion
After upgrading to the release of Weave Gitops that drops support for Flux Beta, Weave Gitops Enterprise will create `v1` Resources.
```

As of writing they do not yet support the new version so please wait before upgrading to Flux GA.
:::

Below, we'll take you through the multiple steps required to migrate to your system to Flux GA. After each step the cluster will be
in a working state, so you can take your time to complete the migration.

1. Upgrade to [latest Weave Gitops](https://docs.gitops.weave.works/docs/releases/)
2. Upgrade to Flux GA in `ClusterBootstrapConfig`s
3. Upgrade to Flux GA on your leaf clusters and management clusters
4. Upgrade GitopsTemplates, GitopsSets and ClusterBootstrapConfigs

### 1. Upgrade to latest WGE

Use your regular WGE upgrade procedure to bring it to the [latest](https://docs.gitops.weave.works/docs/releases/)

:::warning upgrading to Flux GA with WGE v0.22.0?

Ensure that you address [this](#v0220)
:::

At this stage you have Weave Gitops running Flux Beta.

### 2. Upgrade to Flux GA in `ClusterBootstrapConfig`s

First we ensure any new clusters are bootstrapped with Flux GA, then we'll upgrade the existing clusters.

`ClusterBootstrapConfig` will most often contain an invocation of `flux bootstrap`, make sure the image is using `v2`.

> ```patch

Reviewer 3: Could we trim this diff down a bit?
